framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,581 @@
+	dq_init(&g_recording_q);
+	sem_init(&g_sem_recording, 0, 0);
+	pthread_mutex_init(&g_mutex_recording_q, &mutexAttr);
+#endif
+	// Init threads
+	if (pthread_attr_init(&attr) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+
+	sparam.sched_priority = 150;
+	if (pthread_attr_setschedparam(&attr, &sparam) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+
+	if (pthread_attr_setschedpolicy(&attr, SCHED_RR) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+
+	if (pthread_attr_setstacksize(&attr, 8192) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+
+	if (pthread_create(&g_pth_playing, &attr, (pthread_startroutine_t)audio_playing, (void *)NULL) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	if (pthread_create(&g_pth_recording, &attr, (pthread_startroutine_t)audio_recording, (void *)NULL) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+#endif
+	return MEDIA_OK;
+}
+
+void media_shutdown(void)
+{
+	// Shutdown playback thread
+	g_playing_live = false;
+	sem_post(&g_sem_playing);
+	pthread_join(g_pth_playing, NULL);
+	pthread_mutex_destroy(&g_mutex_playing_q);
+	sem_destroy(&g_sem_playing);
+
+	// Shutdown record thread
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	g_recording_live = false;
+	sem_post(&g_sem_recording);
+	pthread_join(g_pth_recording, NULL);
+	pthread_mutex_destroy(&g_mutex_recording_q);
+	sem_destroy(&g_sem_recording);
+#endif
+}
+
+media_result_t media_play(media_t *m, bool loop)
+{
+	pthread_mutex_lock(&g_mutex_playing_q);
+	dq_addlast((dq_entry_t *)m, &g_playing_q);
+	pthread_mutex_unlock(&g_mutex_playing_q);
+	sem_post(&g_sem_playing);
+	return MEDIA_OK;
+}
+
+media_result_t media_record(media_t *m)
+{
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	pthread_mutex_lock(&g_mutex_recording_q);
+	dq_addlast((dq_entry_t *)m, &g_recording_q);
+	pthread_mutex_unlock(&g_mutex_recording_q);
+	sem_post(&g_sem_recording);
+#else
+	pthread_mutex_lock(&g_mutex_playing_q);
+	dq_addlast((dq_entry_t *)m, &g_playing_q);
+	pthread_mutex_unlock(&g_mutex_playing_q);
+	sem_post(&g_sem_playing);
+#endif
+	return MEDIA_OK;
+}
+
+media_result_t media_stop(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PLAYING || m->state == MEDIA_STATE_RECORDING) {
+		m->state = MEDIA_STATE_STOPPING;
+		return MEDIA_OK;
+	}  
+	return MEDIA_ERROR;
+}
+
+media_result_t media_pause(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PLAYING) {
+		m->state = MEDIA_STATE_PAUSING;
+		return MEDIA_OK;
+	}
+	return MEDIA_ERROR;
+}
+
+media_result_t media_resume(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PAUSED) {
+		m->state = MEDIA_STATE_PLAYING;
+		sem_post(&g_sem_playing);
+		return MEDIA_OK;
+	}
+	return MEDIA_ERROR;
+}
+
+void media_setvol(media_t *m, unsigned char vol)
+{
+	return;
+}
+
+bool media_isplaying(media_t *m)
+{
+	return (m->state == MEDIA_STATE_PLAYING);
+}
+
+media_result_t media_getvol(media_t *m)
+{
+	return MEDIA_ERROR;
+}
+
+media_result_t media_seek(media_t *m, int time_ms)
+{
+	return MEDIA_OK;
+}
+
+media_result_t media_setposition(media_t *m, int time_ms)
```

Comment:
below two api's type should be changed.
Also If it is not implemented yet, please remove them from header file.
(If there's any other api which is not implemented yet in header file, please remove.)

Reply:
Ok removed all unimplemented functions